File: internal/database/pipelines.go
// Package database provides PostgreSQL connection pool management and query functions.
package database

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// Pipeline represents a pipeline stored in the database.
type Pipeline struct {
	ID          uuid.UUID
	Name        string
	Description *string
	SatelliteID *uuid.UUID
	CreatedBy   uuid.UUID
	OnFailure   string
	MaxRetries  int
	Schedule    *string
	IsEnabled   bool
	CreatedAt   time.Time
	UpdatedAt   time.Time
	Steps       []PipelineStep
}

// PipelineStep represents a step within a pipeline.
type PipelineStep struct {
	ID         uuid.UUID
	PipelineID uuid.UUID
	StepOrder  int
	AgentID    uuid.UUID
	InputMode  string
	OutputMode string
	Config     string
	CreatedAt  time.Time
}

// PipelineRun represents a pipeline execution run.
type PipelineRun struct {
	ID            uuid.UUID
	PipelineID    uuid.UUID
	SatelliteID   *uuid.UUID
	Status        string
	CurrentStep   *int
	TriggerSource string
	StartedAt     time.Time
	EndedAt       *time.Time
	Error         *string
	CreatedAt     time.Time
	StepRuns      []PipelineStepRun
}

// PipelineStepRun represents a step execution within a pipeline run.
type PipelineStepRun struct {
	ID           uuid.UUID
	PipelineRunID uuid.UUID
	StepOrder    int
	AgentRunID   *uuid.UUID
	Status       string
	InputText    *string
	OutputText   *string
	StartedAt    *time.Time
	EndedAt      *time.Time
	Error        *string
	CreatedAt    time.Time
}

// CreatePipeline inserts a new pipeline into the database.
// Returns the created Pipeline with generated ID or an error.
func CreatePipeline(ctx context.Context, dbPool *pgxpool.Pool, pipeline *Pipeline) error {
	query := `
		INSERT INTO pipelines (
			name, description, satellite_id, created_by, on_failure, max_retries, schedule, is_enabled
		) VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
		RETURNING id, created_at, updated_at
	`

	err := dbPool.QueryRow(ctx, query,
		pipeline.Name,
		pipeline.Description,
		pipeline.SatelliteID,
		pipeline.CreatedBy,
		pipeline.OnFailure,
		pipeline.MaxRetries,
		pipeline.Schedule,
		pipeline.IsEnabled,
	).Scan(
		&pipeline.ID,
		&pipeline.CreatedAt,
		&pipeline.UpdatedAt,
	)
	if err != nil {
		return err
	}
	return nil
}

// GetPipeline retrieves a pipeline by its ID, including all steps sorted by step_order.
// Returns the Pipeline or nil if not found, or an error.
func GetPipeline(ctx context.Context, dbPool *pgxpool.Pool, id uuid.UUID) (*Pipeline, error) {
	// First get the pipeline
	pipelineQuery := `
		SELECT id, name, description, satellite_id, created_by, on_failure, max_retries, schedule, is_enabled, created_at, updated_at
		FROM pipelines
		WHERE id = $1
	`

	var p Pipeline
	var description *string
	var satelliteID *uuid.UUID

	err := dbPool.QueryRow(ctx, pipelineQuery, id).Scan(
		&p.ID,
		&p.Name,
		&description,
		&satelliteID,
		&p.CreatedBy,
		&p.OnFailure,
		&p.MaxRetries,
		&p.Schedule,
		&p.IsEnabled,
		&p.CreatedAt,
		&p.UpdatedAt,
	)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, err
	}
	p.Description = description
	p.SatelliteID = satelliteID

	// Now get all steps sorted by step_order
	stepsQuery := `
		SELECT id, pipeline_id, step_order, agent_id, input_mode, output_mode, config, created_at
		FROM pipeline_steps
		WHERE pipeline_id = $1
		ORDER BY step_order ASC
	`

	rows, err := dbPool.Query(ctx, stepsQuery, id)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var steps []PipelineStep
	for rows.Next() {
		var step PipelineStep
		err := rows.Scan(
			&step.ID,
			&step.PipelineID,
			&step.StepOrder,
			&step.AgentID,
			&step.InputMode,
			&step.OutputMode,
			&step.Config,
			&step.CreatedAt,
		)
		if err != nil {
			return nil, err
		}
		steps = append(steps, step)
	}
	p.Steps = steps

	return &p, rows.Err()
}

// ListPipelines retrieves pipelines with pagination and optional satellite filter.
// Returns a slice of Pipelines, total count, or an error.
func ListPipelines(ctx context.Context, dbPool *pgxpool.Pool, limit, offset int, satelliteID *uuid.UUID) ([]*Pipeline, int, error) {
	// Build dynamic query
	baseQuery := `
		SELECT id, name, description, satellite_id, created_by, on_failure, max_retries, schedule, is_enabled, created_at, updated_at
		FROM pipelines
		WHERE 1=1
	`

	countQuery := `SELECT COUNT(*) FROM pipelines WHERE 1=1`

	args := []interface{}{}
	argIndex := 1

	if satelliteID != nil {
		baseQuery += fmt.Sprintf(" AND satellite_id = $%d", argIndex)
		countQuery += fmt.Sprintf(" AND satellite_id = $%d", argIndex)
		args = append(args, *satelliteID)
		argIndex++
	}

	// Get total count
	var total int
	err := dbPool.QueryRow(ctx, countQuery, args...).Scan(&total)
	if err != nil {
		return nil, 0, err
	}

	// Add pagination
	baseQuery += fmt.Sprintf(" ORDER BY created_at DESC LIMIT $%d OFFSET $%d", argIndex, argIndex+1)
	args = append(args, limit, offset)

	rows, err := dbPool.Query(ctx, baseQuery, args...)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	var pipelines []*Pipeline
	for rows.Next() {
		var p Pipeline
		var description *string
		var satID *uuid.UUID

		err := rows.Scan(
			&p.ID,
			&p.Name,
			&description,
			&satID,
			&p.CreatedBy,
			&p.OnFailure,
			&p.MaxRetries,
			&p.Schedule,
			&p.IsEnabled,
			&p.CreatedAt,
			&p.UpdatedAt,
		)
		if err != nil {
			return nil, 0, err
		}
		p.Description = description
		p.SatelliteID = satID
		pipelines = append(pipelines, &p)
	}

	return pipelines, total, rows.Err()
}

// UpdatePipeline updates a pipeline's mutable fields.
// Returns an error if the update fails.
func UpdatePipeline(ctx context.Context, dbPool *pgxpool.Pool, pipeline *Pipeline) error {
	query := `
		UPDATE pipelines
		SET name = $1, description = $2, satellite_id = $3, on_failure = $4, max_retries = $5, schedule = $6, is_enabled = $7, updated_at = NOW()
		WHERE id = $8
		RETURNING created_at, updated_at
	`

	err := dbPool.QueryRow(ctx, query,
		pipeline.Name,
		pipeline.Description,
		pipeline.SatelliteID,
		pipeline.OnFailure,
		pipeline.MaxRetries,
		pipeline.Schedule,
		pipeline.IsEnabled,
		pipeline.ID,
	).Scan(
		&pipeline.CreatedAt,
		&pipeline.UpdatedAt,
	)
	if err != nil {
		return err
	}
	return nil
}

// DeletePipeline deletes a pipeline by its ID.
// Cascade handles children (pipeline_steps, pipeline_runs, pipeline_step_runs).
// Returns an error if the deletion fails.
func DeletePipeline(ctx context.Context, dbPool *pgxpool.Pool, id uuid.UUID) error {
	_, err := dbPool.Exec(ctx, "DELETE FROM pipelines WHERE id = $1", id)
	return err
}

// CreatePipelineSteps bulk inserts steps for a pipeline.
// Returns an error if the insert fails.
func CreatePipelineSteps(ctx context.Context, dbPool *pgxpool.Pool, pipelineID uuid.UUID, steps []PipelineStep) error {
	if len(steps) == 0 {
		return nil
	}

	query := `
		INSERT INTO pipeline_steps (pipeline_id, step_order, agent_id, input_mode, output_mode, config)
		VALUES ($1, $2, $3, $4, $5, $6)
	`

	for i := range steps {
		_, err := dbPool.Exec(ctx, query,
			pipelineID,
			steps[i].StepOrder,
			steps[i].AgentID,
			steps[i].InputMode,
			steps[i].OutputMode,
			steps[i].Config,
		)
		if err != nil {
			return err
		}
	}

	return nil
}

// DeletePipelineSteps deletes all steps for a pipeline.
// Returns an error if the deletion fails.
func DeletePipelineSteps(ctx context.Context, dbPool *pgxpool.Pool, pipelineID uuid.UUID) error {
	_, err := dbPool.Exec(ctx, "DELETE FROM pipeline_steps WHERE pipeline_id = $1", pipelineID)
	return err
}

// CreatePipelineRun inserts a new pipeline run into the database.
// Returns the created PipelineRun with generated ID or an error.
func CreatePipelineRun(ctx context.Context, dbPool *pgxpool.Pool, run *PipelineRun) error {
	query := `
		INSERT INTO pipeline_runs (pipeline_id, satellite_id, status, current_step, trigger_source, started_at)
		VALUES ($1, $2, $3, $4, $5, NOW())
		RETURNING id, started_at, created_at
	`

	err := dbPool.QueryRow(ctx, query,
		run.PipelineID,
		run.SatelliteID,
		run.Status,
		run.CurrentStep,
		run.TriggerSource,
	).Scan(
		&run.ID,
		&run.StartedAt,
		&run.CreatedAt,
	)
	if err != nil {
		return err
	}
	return nil
}

// GetPipelineRun retrieves a pipeline run by its ID, including all step runs sorted by step_order.
// Returns the PipelineRun or nil if not found, or an error.
func GetPipelineRun(ctx context.Context, dbPool *pgxpool.Pool, id uuid.UUID) (*PipelineRun, error) {
	// First get the pipeline run
	runQuery := `
		SELECT id, pipeline_id, satellite_id, status, current_step, trigger_source, started_at, ended_at, error, created_at
		FROM pipeline_runs
		WHERE id = $1
	`

	var run PipelineRun
	var satelliteID *uuid.UUID
	var currentStep *int
	var endedAt *time.Time
	var error *string

	err := dbPool.QueryRow(ctx, runQuery, id).Scan(
		&run.ID,
		&run.PipelineID,
		&satelliteID,
		&run.Status,
		&currentStep,
		&run.TriggerSource,
		&run.StartedAt,
		&endedAt,
		&error,
		&run.CreatedAt,
	)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, err
	}
	run.SatelliteID = satelliteID
	run.CurrentStep = currentStep
	run.EndedAt = endedAt
	run.Error = error

	// Now get all step runs sorted by step_order
	stepRunsQuery := `
		SELECT id, pipeline_run_id, step_order, agent_run_id, status, input_text, output_text, started_at, ended_at, error, created_at
		FROM pipeline_step_runs
		WHERE pipeline_run_id = $1
		ORDER BY step_order ASC
	`

	rows, err := dbPool.Query(ctx, stepRunsQuery, id)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var stepRuns []PipelineStepRun
	for rows.Next() {
		var stepRun PipelineStepRun
		var agentRunID *uuid.UUID
		var inputText *string
		var outputText *string
		var startedAt *time.Time
		var endedAtStep *time.Time
		var errorStep *string

		err := rows.Scan(
			&stepRun.ID,
			&stepRun.PipelineRunID,
			&stepRun.StepOrder,
			&agentRunID,
			&stepRun.Status,
			&inputText,
			&outputText,
			&startedAt,
			&endedAtStep,
			&errorStep,
			&stepRun.CreatedAt,
		)
		if err != nil {
			return nil, err
		}
		stepRun.AgentRunID = agentRunID
		stepRun.InputText = inputText
		stepRun.OutputText = outputText
		stepRun.StartedAt = startedAt
		stepRun.EndedAt = endedAtStep
		stepRun.Error = errorStep
		stepRuns = append(stepRuns, stepRun)
	}
	run.StepRuns = stepRuns

	return &run, rows.Err()
}

// UpdatePipelineRun updates a pipeline run's fields (status, current_step, ended_at, error).
// Returns an error if the update fails.
func UpdatePipelineRun(ctx context.Context, dbPool *pgxpool.Pool, run *PipelineRun) error {
	query := `
		UPDATE pipeline_runs
		SET status = $1, current_step = $2, ended_at = $3, error = $4
		WHERE id = $5
		RETURNING started_at, created_at
	`

	err := dbPool.QueryRow(ctx, query,
		run.Status,
		run.CurrentStep,
		run.EndedAt,
		run.Error,
		run.ID,
	).Scan(
		&run.StartedAt,
		&run.CreatedAt,
	)
	if err != nil {
		return err
	}
	return nil
}

// ListPipelineRuns retrieves pipeline runs for a specific pipeline with pagination.
// Returns a slice of PipelineRuns, total count, or an error.
func ListPipelineRuns(ctx context.Context, dbPool *pgxpool.Pool, pipelineID uuid.UUID, limit, offset int) ([]*PipelineRun, int, error) {
	// Get total count
	var total int
	countQuery := `SELECT COUNT(*) FROM pipeline_runs WHERE pipeline_id = $1`
	err := dbPool.QueryRow(ctx, countQuery, pipelineID).Scan(&total)
	if err != nil {
		return nil, 0, err
	}

	// Get paginated results
	query := `
		SELECT id, pipeline_id, satellite_id, status, current_step, trigger_source, started_at, ended_at, error, created_at
		FROM pipeline_runs
		WHERE pipeline_id = $1
		ORDER BY created_at DESC
		LIMIT $2 OFFSET $3
	`

	rows, err := dbPool.Query(ctx, query, pipelineID, limit, offset)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	var runs []*PipelineRun
	for rows.Next() {
		var run PipelineRun
		var satelliteID *uuid.UUID
		var currentStep *int
		var endedAt *time.Time
		var error *string

		err := rows.Scan(
			&run.ID,
			&run.PipelineID,
			&satelliteID,
			&run.Status,
			&currentStep,
			&run.TriggerSource,
			&run.StartedAt,
			&endedAt,
			&error,
			&run.CreatedAt,
		)
		if err != nil {
			return nil, 0, err
		}
		run.SatelliteID = satelliteID
		run.CurrentStep = currentStep
		run.EndedAt = endedAt
		run.Error = error
		runs = append(runs, &run)
	}

	return runs, total, rows.Err()
}

// CreatePipelineStepRun inserts a new pipeline step run into the database.
// Returns the created PipelineStepRun with generated ID or an error.
func CreatePipelineStepRun(ctx context.Context, dbPool *pgxpool.Pool, stepRun *PipelineStepRun) error {
	query := `
		INSERT INTO pipeline_step_runs (pipeline_run_id, step_order, status)
		VALUES ($1, $2, $3)
		RETURNING id, created_at
	`

	err := dbPool.QueryRow(ctx, query,
		stepRun.PipelineRunID,
		stepRun.StepOrder,
		stepRun.Status,
	).Scan(
		&stepRun.ID,
		&stepRun.CreatedAt,
	)
	if err != nil {
		return err
	}
	return nil
}

// UpdatePipelineStepRun updates a pipeline step run's fields (status, agent_run_id, input_text, output_text, started_at, ended_at, error).
// Returns an error if the update fails.
func UpdatePipelineStepRun(ctx context.Context, dbPool *pgxpool.Pool, stepRun *PipelineStepRun) error {
	query := `
		UPDATE pipeline_step_runs
		SET status = $1, agent_run_id = $2, input_text = $3, output_text = $4, started_at = $5, ended_at = $6, error = $7
		WHERE id = $8
		RETURNING created_at
	`

	err := dbPool.QueryRow(ctx, query,
		stepRun.Status,
		stepRun.AgentRunID,
		stepRun.InputText,
		stepRun.OutputText,
		stepRun.StartedAt,
		stepRun.EndedAt,
		stepRun.Error,
		stepRun.ID,
	).Scan(
		&stepRun.CreatedAt,
	)
	if err != nil {
		return err
	}
	return nil
}

File: cockpit/src/api/pipelines.ts
/**
 * Pipeline API Client
 * 
 * TypeScript types and API functions for pipeline operations.
 * Provides functions for creating, managing, and running pipelines.
 */

import { apiRequest } from './client';

// ============================================================================
// Types
// ============================================================================

/**
 * Pipeline step configuration
 */
export interface PipelineStep {
  id: string;
  pipeline_id: string;
  step_order: number;
  agent_id: string;
  input_mode: 'none' | 'previous' | 'manual';
  output_mode: 'none' | 'pass' | 'collect';
  config: Record<string, unknown>;
}

/**
 * Pipeline step run status
 */
export interface PipelineStepRun {
  step_id: string;
  step_order: number;
  agent_run_id?: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'skipped';
  started_at?: string;
  ended_at?: string;
  output?: string;
  error?: string;
}

/**
 * Pipeline run status
 */
export type PipelineRunStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

/**
 * Pipeline trigger source
 */
export type TriggerSource = 'manual' | 'schedule' | 'webhook';

/**
 * Pipeline run
 */
export interface PipelineRun {
  id: string;
  pipeline_id: string;
  satellite_id: string;
  status: PipelineRunStatus;
  current_step: number;
  trigger_source: TriggerSource;
  started_at: string;
  ended_at?: string;
  error?: string;
  step_runs: PipelineStepRun[];
}

/**
 * Pipeline model
 */
export interface Pipeline {
  id: string;
  name: string;
  description: string;
  satellite_id: string;
  on_failure: 'abort' | 'continue' | 'retry';
  max_retries: number;
  schedule?: string;
  is_enabled: boolean;
  steps: PipelineStep[];
  created_at: string;
  updated_at: string;
}

// ============================================================================
// Request Types
// ============================================================================

/**
 * Create pipeline request payload
 */
export interface CreatePipelineRequest {
  name: string;
  description?: string;
  satellite_id: string;
  on_failure?: 'abort' | 'continue' | 'retry';
  max_retries?: number;
  schedule?: string;
  is_enabled?: boolean;
  steps?: Omit<PipelineStep, 'id' | 'pipeline_id'>[];
}

/**
 * Update pipeline request payload
 */
export interface UpdatePipelineRequest {
  name?: string;
  description?: string;
  satellite_id?: string;
  on_failure?: 'abort' | 'continue' | 'retry';
  max_retries?: number;
  schedule?: string;
  is_enabled?: boolean;
  steps?: Omit<PipelineStep, 'id' | 'pipeline_id'>[];
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Create a new pipeline
 * 
 * POST /api/v1/pipelines
 */
export async function createPipeline(req: CreatePipelineRequest): Promise<Pipeline> {
  return apiRequest<Pipeline>('/pipelines', {
    method: 'POST',
    body: JSON.stringify(req),
  });
}

/**
 * Get a pipeline by ID
 * 
 * GET /api/v1/pipelines/:id
 */
export async function getPipeline(id: string): Promise<Pipeline> {
  return apiRequest<Pipeline>(`/pipelines/${id}`);
}

/**
 * List pipelines with optional filtering and pagination
 * 
 * GET /api/v1/pipelines
 */
export interface ListPipelinesParams {
  limit?: number;
  offset?: number;
  satellite_id?: string;
}

export async function listPipelines(params?: ListPipelinesParams): Promise<{ pipelines: Pipeline[], total: number }> {
  const queryParams = new URLSearchParams();

  if (params?.limit !== undefined) {
    queryParams.append('limit', params.limit.toString());
  }
  if (params?.offset !== undefined) {
    queryParams.append('offset', params.offset.toString());
  }
  if (params?.satellite_id) {
    queryParams.append('satellite_id', params.satellite_id);
  }

  const queryString = queryParams.toString();
  const endpoint = `/pipelines${queryString ? `?${queryString}` : ''}`;

  return apiRequest<{ pipelines: Pipeline[], total: number }>(endpoint);
}

/**
 * Update a pipeline
 * 
 * PUT /api/v1/pipelines/:id
 */
export async function updatePipeline(id: string, req: UpdatePipelineRequest): Promise<Pipeline> {
  return apiRequest<Pipeline>(`/pipelines/${id}`, {
    method: 'PUT',
    body: JSON.stringify(req),
  });
}

/**
 * Delete a pipeline
 * 
 * DELETE /api/v1/pipelines/:id
 */
export async function deletePipeline(id: string): Promise<void> {
  await apiRequest<{ status: string }>(`/pipelines/${id}`, {
    method: 'DELETE',
  });
}

/**
 * Trigger a pipeline run
 * 
 * POST /api/v1/pipelines/:id/run
 */
export async function runPipeline(id: string, satelliteId?: string): Promise<{ run_id: string }> {
  const body = satelliteId ? { satellite_id: satelliteId } : {};
  return apiRequest<{ run_id: string }>(`/pipelines/${id}/run`, {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

/**
 * List pipeline runs with pagination
 * 
 * GET /api/v1/pipelines/:pipelineId/runs
 */
export interface ListPipelineRunsParams {
  limit?: number;
  offset?: number;
}

export async function listPipelineRuns(pipelineId: string, params?: ListPipelineRunsParams): Promise<{ runs: PipelineRun[], total: number }> {
  const queryParams = new URLSearchParams();

  if (params?.limit !== undefined) {
    queryParams.append('limit', params.limit.toString());
  }
  if (params?.offset !== undefined) {
    queryParams.append('offset', params.offset.toString());
  }

  const queryString = queryParams.toString();
  const endpoint = `/pipelines/${pipelineId}/runs${queryString ? `?${queryString}` : ''}`;

  return apiRequest<{ runs: PipelineRun[], total: number }>(endpoint);
}

/**
 * Get a specific pipeline run
 * 
 * GET /api/v1/pipelines/runs/:runId
 */
export async function getPipelineRun(runId: string): Promise<PipelineRun> {
  return apiRequest<PipelineRun>(`/pipelines/runs/${runId}`);
}

/**
 * Set pipeline schedule (cron expression)
 * 
 * POST /api/v1/pipelines/:id/schedule
 */
export async function setPipelineSchedule(id: string, cronExpr: string): Promise<void> {
  await apiRequest<{ status: string }>(`/pipelines/${id}/schedule`, {
    method: 'POST',
    body: JSON.stringify({ cron: cronExpr }),
  });
}

/**
 * Delete pipeline schedule
 * 
 * DELETE /api/v1/pipelines/:id/schedule
 */
export async function deletePipelineSchedule(id: string): Promise<void> {
  await apiRequest<{ status: string }>(`/pipelines/${id}/schedule`, {
    method: 'DELETE',
  });
}

// ============================================================================
// Exports
// ============================================================================

export default {
  createPipeline,
  getPipeline,
  listPipelines,
  updatePipeline,
  deletePipeline,
  runPipeline,
  listPipelineRuns,
  getPipelineRun,
  setPipelineSchedule,
  deletePipelineSchedule,
};
